### PR TITLE
Improve readability of format strings

### PR DESCRIPTION
**Pull request by thisch**
cc @phikal

### CODE REVIEW

**Comment by phikal** [comment]
I see you've pinged me, but it's not really up to me to say anything. All I can offer is, as a general elisp point,  to point out that the following expressions all evaluate to `t`:

~~~elisp
(string= "two\nlines"
	 "two
lines")
~~~

~~~elisp
(string= "two\nlines"
	 "\
two
lines")
~~~

and that it might be cleaner to use this instead of `mapconcat`.

**Comment by thisch** [comment]
Sry for pinging you. I like you suggestion! Thx for mentioning it. I've just pushed my new changes.

**Comment by phikal** [comment]
Don't worry, I'm just glad to help :)

**Comment by jacktasia** [comment]
I am not sure this is worth the increased vertical space in the code. Since this is not actually changing the output itself. If we're going to make a change then a list of tuples where the title string and variable with the value are combined would probably be a bigger improvement then we can just have a single template string and iterate through the list of the tuples. thoughts?

**Comment by thisch** [comment]
WDYT about this helper defun that builds the args passed to `dumb-jump-message`?

```el

(defun dumb-jump-debug-message (defunname pairs)
  "...."
  (let ((args (seq-map #'cdr pairs))
        (fmtmiddle (mapconcat (lambda (param)
                                (format "%s:\n\t%%s\n" (car param))) pairs "\n")))
    (apply #'dumb-jump-message
           `(,(concat "DUMB JUMP DEBUG `"
                      defunname
                      "` START\n----\n\n"
                      fmtmiddle
                      "\n-----\nDUMB JUMP DEBUG `"
                      defunname
                      "` END\n-----")
             ,@args))))


(dumb-jump-debug-message "defunname" '(("varname1" .  "value1")
                                       ("varname2" . "value2")))
```

**Comment by phikal** [comment]
If you were to look into something like this, I think a signature like

        (defmacro dumb-jump-debug-message (&rest pairs)
          ...)

assuming one would find a way to get the name of the function in which
the macro is being expanded. That would avoid the duplicate "defunname"
argument. 

Perhaps this answer could help https://emacs.stackexchange.com/a/2312?

-- 
	Philip K.

**Comment by phikal** [comment]
Oh and

~~~
(apply #'dumb-jump-message
           `(,(concat "DUMB JUMP DEBUG `"
                      defunname
                      "` START\n----\n\n"
                      fmtmiddle
                      "\n-----\nDUMB JUMP DEBUG `"
                      defunname
                      "` END\n-----")
             ,@args))
~~~

is equivalent to 

~~~
(apply #'dumb-jump-message
       (concat "DUMB JUMP DEBUG `"
                  defunname
                  "` START\n----\n\n"
                  fmtmiddle
                  "\n-----\nDUMB JUMP DEBUG `"
                  defunname
                  "` END\n-----")
       arg)
~~~

**Comment by thisch** [comment]
Thx for the tips @phikal ! I have a macro know that should work:

```el
(defmacro dumb-jump-debug-macro (defunname &rest pairs)
  "...."
  (let ((args (seq-map #'cdr pairs))
        (fmtmiddle (mapconcat (lambda (param)
                                (format "%s:\n\t%%s\n" (car param))) pairs "\n")))
    `(apply #'format
            (concat "DUMB JUMP DEBUG `"
                    ,defunname
                    "` START\n----\n\n"
                    ,fmtmiddle
                    "\n-----\nDUMB JUMP DEBUG `"
                    ,defunname
                    "` END\n-----")
            (quote ,args))))

(macroexpand-all (dumb-jump-debug-macro "defunname"
                                        ("varname1" . "value1")
                                        ("varname2" . "value2")))
"DUMB JUMP DEBUG `defunname` START
----

varname1:
	value1

varname2:
	value2

-----
DUMB JUMP DEBUG `defunname` END
-----"
```

I haven't had a look at the removal of the defunname arg.

**Comment by phikal** [comment]
I was thinking more of something along these lines

~~~el
(defmacro dumb-jump-debug-message (&rest vars)
  (let ((frames) (frame) (i 5) name)
    (while (setq frame (backtrace-frame i))
      (push frame frames)
      (cl-incf i))
    (setq name (thread-last (reverse frames)
		 (cl-find-if
		  (lambda (frame)
		    (ignore-errors (and (car frame)
					(eq (caaddr frame) 'defalias)))))
		 cl-third cadadr symbol-name))
    (with-temp-buffer
      (insert "DUMB JUMP DEBUG `")
      (insert name)
      (insert "` START\n----\n\n")
      (dolist (var vars)
	(insert (symbol-name var) ":\n\t%s\n\n"))
      (insert "\n-----\nDUMB JUMP DEBUG `")
      (insert name)
      (insert "` END\n-----")
      `(dumb-jump-message
	,(buffer-string)
	,@(mapcar (lambda (var)
		    `(eval ',var))
		  vars)))))
~~~

But again, I'm not sure how useful this is, considering that it appears only a few times in the current codebase.

File: dumb-jump.el [hunk 1]
```diff
@@ -1844,6 +1845,35 @@ Optionally pass t for RUN-NOT-TESTS to see a list of all failed rules"
...
 
+(defmacro dumb-jump-debug-message (&rest exprs)
+  "Generate a debug message to print all expressions EXPRS."
+  (declare (indent defun))
+  (let ((i 5) frames frame defunname)
+    ;; based on https://emacs.stackexchange.com/a/2312
+    (while (setq frame (backtrace-frame i))
+      (push frame frames)
+      (cl-incf i))
+    (setq defunname (thread-last (reverse frames)
```
[the remark this answers is not in the document]
Comment (thisch) [reply dumb-jump.el:1856]: The expanded version of it looks like

```
(setq defunname (symbol-name
 (cl-cadadr
  (cl-caddr
   (cl-find-if
    (lambda
      (frame)
      (ignore-errors
        (and
         (car frame)
         (eq
          (cl-caaddr frame)
          'defalias))))
    (reverse frames))))))
```

**Comment by thisch** [comment]
Thx a lot for your help. The reason why I'm motivated to work on it is that I can probably also use this macro in my elisp packages.

**Comment by phikal** [comment]
I think you have to update the commit message, it still mentions Emacs 25.

Also, by convention, it should be `defun-name` not `defunname`.

**Comment by thisch** [comment]
> I think you have to update the commit message, it still mentions Emacs 25.
> 
> Also, by convention, it should be `defun-name` not `defunname`.

done

**Pull request merged**